numpy/core/include/numpy/ndarraytypes.h

```diff
@@ -1027,6 +1027,8 @@ typedef void (NpyIter_GetMultiIndexFunc)(NpyIter *iter,
  * eliminate overlap.
  */
 #define NPY_ITER_COPY_IF_OVERLAP            0x00002000
+/* At least one operand uses UPDATEIFCOPY semantics, use context manager */
+#define NPY_ITER_USE_CONTEXT_MANAGER        0x00004000
```

Comment:
I will add "updating cython/docs/src/userguide/pypy.rst" to my TODO list. While not wrong, that page is slightly out of date and it should mention that only certain C-API functions will actually create a python-level object linked to the C one. Looking at the [implementation of ``Py_DECREF``](https://bitbucket.org/pypy/pypy/src/0b72fd1a7641318e86fbc1134f4f23ab632c7484/pypy/module/cpyext/pyobject.py?at=default&fileviewer=file-view-default#pyobject.py-315) one can see that if the refcount is 0 (which means no python-level object is created, otherwise the refcount would be >= ``REFCNT_FROM_PYPY``) then ``obj.tp_dealloc``is called immediately.

I did try to naively add a call to ``PyArray_ResolveUpdateIfCopy`` inside ``NpyIter_Deallocate`` just before the call to ``Py_XDECREF(*object)``, but it causes a crash in the ``test_nditer.py`` tests. It seems there are cases where the operand is not solely owned by the iterator, i.e. in any of the places ``NpyIter_GetOperandArray``, ``NpyIter_Copy`` or others are called.